File: monitor.py
# monitor.py
import click
import json
from kafka import KafkaConsumer
from rich.live import Live
from rich.table import Table
from datetime import datetime

def generate_table(data_rows: list) -> Table:
    """Generates a Rich table from a list of recent messages."""
    table = Table(title="VibeFactor Engine: Real-Time Sentiment Monitor")
    table.add_column("Timestamp", style="cyan", no_wrap=True)
    table.add_column("Text", style="magenta", width=80)
    table.add_column("Compound Score", justify="right", style="green")

    for row in data_rows:
        ts = datetime.fromtimestamp(row['timestamp']).strftime('%Y-%m-%d %H:%M:%S')
        score = row['sentiment']['compound']
        
        # Color-code the score for quick visual scanning
        color = "bold green" if score > 0.05 else "bold red" if score < -0.05 else "yellow"
        score_str = f"[{color}]{score:+.4f}[/{color}]"
        
        table.add_row(ts, row['text'], score_str)
        
    return table

@click.command()
@click.option('--topic', default='sentiment_stream', help='The Kafka topic to monitor.')
@click.option('--broker', default='localhost:9092', help='The Kafka broker address.')
def monitor(topic: str, broker: str):
    """
    A real-time CLI monitor for the sentiment analysis pipeline,
    built with Rich and Click.
    """
    consumer = KafkaConsumer(
        topic,
        bootstrap_servers=broker,
        value_deserializer=lambda v: json.loads(v.decode('utf-8')),
        auto_offset_reset='latest' # Only show new messages
    )
    print("⏳ Connecting to Kafka stream for monitoring...")
    
    recent_messages = []
    max_rows = 15 # Number of rows to display in the terminal

    with Live(generate_table(recent_messages), refresh_per_second=4, screen=True) as live:
        live.console.print("✅ Connected. Waiting for sentiment signals...")
        for message in consumer:
            recent_messages.append(message.value)
            if len(recent_messages) > max_rows:
                recent_messages.pop(0)
            live.update(generate_table(recent_messages))

if __name__ == '__main__':
    monitor()

File: monitor.cpp
#include <iostream>
#include <string>
#include <iomanip>
#include <librdkafka/rdkafkacpp.h>
#include <chrono>
#include <ctime>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

int main() {
    std::string errstr;
    
    RdKafka::Conf *conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
    conf->set("bootstrap.servers", "localhost:9092", errstr);
    conf->set("group.id", "monitor_cpp_group", errstr);
    conf->set("auto.offset.reset", "latest", errstr);

    RdKafka::KafkaConsumer *consumer = RdKafka::KafkaConsumer::create(conf, errstr);
    if (!consumer) { std::cerr << "Failed to create consumer: " << errstr << std::endl; return 1; }
    consumer->subscribe({"sentiment_stream"});

    std::cout << "⏳ Connecting to Kafka stream for monitoring..." << std::endl;
    std::cout << "✅ Connected. Waiting for sentiment signals...\n" << std::endl;
    
    // Table Header
    std::cout << std::left << std::setw(25) << "Timestamp" << std::setw(80) << "Text" << "Score" << std::endl;
    std::cout << std::string(115, '-') << std::endl;

    while (true) {
        RdKafka::Message *msg = consumer->consume(1000);
        if (msg->err() == RdKafka::ERR_NO_ERROR) {
            std::string payload(static_cast<const char *>(msg->payload()), msg->len());
            try {
                json data = json::parse(payload);
                double ts = data["timestamp"];
                std::string text = data["text"];
                double score = data["sentiment"]["compound"];

                // Format Timestamp
                std::time_t t = static_cast<std::time_t>(ts);
                char buf[30];
                std::strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", std::localtime(&t));

                // Terminal ANSI Color Codes
                std::string color = "\033[0m"; // Default Reset
                if (score > 0.05) color = "\033[1;32m"; // Bold Green
                else if (score < -0.05) color = "\033[1;31m"; // Bold Red
                else color = "\033[1;33m"; // Yellow

                std::cout << std::left << std::setw(25) << buf
                          << std::setw(80) << text.substr(0, 75)
                          << color << std::showpos << std::fixed << std::setprecision(4) << score << "\033[0m" << std::noshowpos
                          << std::endl;

            } catch (const std::exception& e) {
                std::cerr << "Parse error: " << e.what() << std::endl;
            }
        }
        delete msg;
    }
    return 0;
}